File: src/audit.c
#include "../include/bcind_kernel.h"

void log_audit_event(const char* filepath, const char* action, const char* status, double snr_db, uint64_t seq) {
    if (filepath == NULL) return;

    FILE* fp = fopen(filepath, "a");
    if (!fp) return;

    time_t now = time(NULL);
    char time_buf[64];
    struct tm* tm_info = gmtime(&now);
    strftime(time_buf, sizeof(time_buf), "%Y-%m-%dT%H:%M:%SZ", tm_info);

    fprintf(fp, "{\"timestamp\":\"%s\",\"sequence\":%lu,\"action\":\"%s\",\"status\":\"%s\",\"snr_db\":%.2f}\n",
            time_buf, (unsigned long)seq, action ? action : "UNKNOWN", status ? status : "UNKNOWN", snr_db);

    fclose(fp);
    printf("[AUDIT] Event sequence %lu appended to %s\n", (unsigned long)seq, filepath);
}

File: src/audit.rs
use chrono::Utc;
use serde::Serialize;
use std::fs::OpenOptions;
use std::io::Write;
use std::path::Path;

#[derive(Serialize)]
struct AuditEvent<'a> {
    timestamp: String,
    sequence: u64,
    action: &'a str,
    status: &'a str,
    snr_db: f64,
}

pub fn log_audit_event<P: AsRef<Path>>(
    filepath: P,
    action: &str,
    status: &str,
    snr_db: f64,
    seq: u64,
) {
    let timestamp = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
    let event = AuditEvent {
        timestamp,
        sequence: seq,
        action,
        status,
        snr_db,
    };

    if let Ok(serialized) = serde_json::to_string(&event) {
        if let Some(parent) = filepath.as_ref().parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        if let Ok(mut file) = OpenOptions::new().create(true).append(true).open(&filepath) {
            let _ = writeln!(file, "{}", serialized);
            println!(
                "[AUDIT] Event sequence {} appended to {}",
                seq,
                filepath.as_ref().display()
            );
        }
    }
}
